### ISSUE TITLE
Missing provider failure case leaves s3 lock in place

### ISSUE THREAD

**Issue by joe-bowman** [hashicorp/terraform#23318]
```
Terraform v0.12.13
+ provider.aws v2.34.0
+ provider.google v2.18.1
+ provider.template v2.1.2
+ provider.vault v2.2.0
```

### Terraform Configuration Files
Attempt to use a Google Cloud resource without a relevant provider being declared

### Expected Behavior
Lock should be removed in event of failure case.

### Actual Behavior
Error on `terraform import` (correct)
```
Error: provider.google-beta: no suitable version installed
  version requirements: "(any version)"
  versions installed: none
```

And next operation:
```
Error: Error locking state: Error acquiring the state lock: ConditionalCheckFailedException: The conditional request failed
	status code: 400, request id: SL3JPK86NVLJITHD2I5E6D9PHJVV4KQNSO5AEMVJF66Q9ASUAAJG
Lock Info:
  ID:        53b7efb8-a7a7-a0c3-f55f-73aee25c8472
  Path:      [redacted].tfstate
  Operation: OperationTypeInvalid
  Who:       joe@Joes-MacBook-Pro.local
  Version:   0.12.13
  Created:   2019-11-08 10:55:32.551892 +0000 UTC
  Info:
```

Lock never gets releases after the import failure.

### Steps to Reproduce
1. Create a state stored in s3
2. Create a new google resource, without a appropriate provider
3. Run (in my case `terraform import` on the google resource.

**Comment by alisdair** [comment]
Thanks for filing this! I was able to confirm this behaviour with 0.12.20. My reproduction steps:

1. Create `main.tf`:

    ```hcl
    resource null_resource example {
      count = 1

      triggers = {
        index = count.index
      }
    }

    # resource example_resource whatever {}

    terraform {
      backend "s3" {
        bucket = "your-bucket"
        key = "remote-state/tf-example"
        region = "us-east-1"
        dynamodb_table = "your-dynamodb-table"
      }
    }
    ```
1. `terraform init`, `terraform apply` to create a state version
1. Uncomment the example resource
1. `terraform import example_resource.whatever abc123` to trigger a missing provider error

This last step fails and leaves a dangling lock.

**Comment by alisdair** [comment]
Thanks for reporting this issue! A fix has been merged and will be part of the next 0.12 release.
